### ISSUE TITLE
Testing that an email is sent

### ISSUE THREAD

**Issue by drhumlen** [playframework/play-mailer#73]
Is there a way to test that an email is sent, and check the contents of it? Would be nice to do something like:

```
"A verification email is sent when the user has registered" in {
  MailPlugin.startCapturing()
  Users.register(name = "Frodo", email = "frodo@example.com")
  val emails: List[Email] = MailPlugin.stopCaptureAndCollectResults()

  emails mustEqual List(
    Email(
      subject = "Welcome Frodo",
      recipient = "frodo@example.com",
      bodyText = "Here is your verifcation link: ______"
    )
}
```

Or perhaps something more "functional"/safe:

```
"A verification email is sent when the user has registered" in {
  val emails: List[Email] = MailPlugin.captureMailSentInBlock {
    Users.register(name = "Frodo", email = "frodo@example.com")
  }

  emails mustEqual List(
    Email(
      subject = "Welcome Frodo",
      recipient = "frodo@example.com",
      bodyText = "Here is your verifcation link: ______"
    )
}
```

Is it possible to do something like this at all right now? ie. Setting the plugin in testing/capture mode so that we can verify: (1) that en email is sent,  (2) verify the contents of it.

Also, mock-mode must be on when running your test suite, right? (`play.mailer.mock = yes`)

**Comment by Mogztter** [comment]
Hello,

If you are using Play 2.4 then you can use dependency injection to provide a mock of MailerClient to your UserModule.

With Guice (runtime dependency injection):
```scala
val application = new GuiceApplicationBuilder()
  .overrides(bind[MailerClient].to[MockMailerClient])
  .build
```
https://www.playframework.com/documentation/2.4.x/ScalaTestingWithGuice

You can create a mock of MailerClient using Mockito or ScalaMock. Then you can assert that `sendEmail` method was called with the expected `Email` object, something like:

```scala
"A verification email is sent when the user has registered" in new WithApplication(application) {
  UserModule.registerNewUser(name = "Frodo", email = "frodo@example.com")
  Mockito.verify(UserModule.mailerClient).sendEmail(Email(
      subject = "Welcome Frodo",
      recipient = "frodo@example.com",
      bodyText = "Here is your verifcation link: ______"
    ))
}
```

> Also, mock-mode must be on when running your test suite, right? (play.mailer.mock = yes)

No `play.mailer.mock` should be considered deprecated because you can now choose to inject a MockMailer with the qualifier "mock", see: https://github.com/playframework/play-mailer/blob/master/src/main/scala/play/api/libs/mailer/MailerPlugin.scala#L25

Also in your test you want to do assertion. The `MockMailer` in play-mailer will just prints out the email data, see: https://github.com/playframework/play-mailer/blob/master/src/main/scala/play/api/libs/mailer/MailerPlugin.scala#L248

> I'm sorry if this is asked before, but I couldn't find anything in the manual or from searching.

Don't apologize this is a very good question and I need to add documentation about testing.
If something is not clear, feel free to ask again.

**Comment by drhumlen** [comment]
Thanks for swift reply :)

Unfortunately, Play's support for dependency injection was a little too late for our project :( . We're already 2.5 years into a quite big play-project, so we don't have time to rewrite everything to use dependency injection.

I would go for the solution you suggested above though when/if we start another play-project, but for now I made a small workaround to work with our existing code:

```
import play.api.Play.current
import play.api.libs.mailer.{Email, MailerPlugin}

object Mailer {
  private var capturedEmails: List[Email] = Nil
  private var captureMode = false

  def send(email: Email): Unit = {
    if (captureMode == true) {
      capturedEmails = email :: capturedEmails
    } else {
      MailerPlugin.send(email)
    }
  }

  def captureMailsSentIn(blk: => Unit): List[Email] = {
    captureMode = true
    capturedEmails = Nil
    blk
    captureMode = false
    capturedEmails
  }
}
```

It's not exactly 100% functional, but it does what must be done in order to test it. If anybody else is in the same situation as us, it's probably a good idea to set `play.mailer.mock=yes` since all code that is not run directly inside `captureMailsSentIn` will still send emails around to god-knows-who.

Anyway, just thought I'd post this here in case someone else has painted themselves into a DI-less-corner like we have. Here's how we use it in one of our tests:

```
    "a login-token is sent on email when registering" in {
      val capturedEmails = Mailer.captureMailsSentIn {
        Helpers.reqTest(
          routes.ApplicantController.registerForeign(),
          input = ForeignRegistrationFormData(
            name = "Frodo",
            birthDate = DateTime.now,
            email = "frodo@example.com",
            mobile = "12345678",
            gender = Gender.Male
          ),
          output = Json.obj("id" -> JsonMatcher.___anyString),
          expectedStatusCode = CREATED
        )
      }

      capturedEmails must haveLength(1)
      val mail = capturedEmails(0)

      mail.to === Seq("frodo@example.com")
      mail.bodyText.get must be containing "http://localhost:9000/api/verify?code="
    }
```

**Comment by Mogztter** [comment]
Hello @drhumlen, thanks for sharing your findings.
I think you can also use [ScalaMock](http://scalamock.org/) because its allow to mock singleton and companion objects. So I think you can do something like:

```scala
val mailerPluginMock = mockObject(MailerPlugin)
UserModule.registerNewUser(name = "Frodo", email = "frodo@example.com")
val expectedEmail = Email(
  subject = "Welcome Frodo",
  recipient = "frodo@example.com",
  bodyText = "Here is your verifcation link: ______"
)
mailerPluginMock.expects.sendEmail(expectedEmail)
```

**Comment by placebocurejava** [comment]
Hi,

Is there any suggestions for PLAYA JAVA to set up veritifcation of sent messages ?
I receive the message : Email sent! but no messages was received in my inbox ...

**Comment by yuchaosydney** [comment]
Sorry, I might want to ask is there any full example to test sending mail function? I am just trying to get start.

**Comment by placebocurejava** [comment]
Hello
Of course i can help with samples and examples
Tell me what do u want exactly please..

**Comment by marcospereira** [comment]
Hey,

The way I would recommend that is to have a separated configuration for your test environment, having just the email settings change, and then use something like [Wiser](https://github.com/voodoodyne/subethasmtp/blob/master/Wiser.md) or [GreenMail](http://www.icegreen.com/greenmail/) which are both made to test email sending.

**Comment by yuchaosydney** [comment]
`val cid = "1234"
    val email = Email(
      "Simple email",
      "Yuchao's Blog <yuchao@email.com>",
      Seq("Miss TO <yuchao.sydneyuni@gmail.com>"),
      // adds attachment
      attachments = Seq(
      ),
      // sends text, HTML or both...
      bodyText = Some("A text message"),
      bodyHtml = Some(s"""<html><body><p>An <b>html</b> message with cid <img src="cid:$cid"></p></body></html>""")
    )
    mailerClient.send(email)`

Just simple this block of code, should be easy to make a test case but cost me whole night.

Appreciate any help to a beginner like me.

Thanks!

**Comment by placebocurejava** [comment]
Hi
Feel free to share ur screen with me tomorrow and we will launch ur sending
mails tomorrow ok ?

**Comment by yuchaosydney** [comment]
Sorry I just reformat my comment here. Code should be readable now.

Yeah no problem, I will just try @marcospereira 's advice at the same time.

Thanks for your attention.


```
     val cid = "1234"
     val email = Email(
      "Simple email",
      "Yuchao's Blog <yuchao@email.com>",
      Seq("Miss TO <yuchao.sydneyuni@gmail.com>"),
      // adds attachment
      attachments = Seq(
      ),
      // sends text, HTML or both...
      bodyText = Some("A text message"),
      bodyHtml = Some(s"""<html><body><p>An <b>html</b> message with cid <img src="cid:$cid"></p></body></html>""")
    )
    mailerClient.send(email)
```

**Comment by placebocurejava** [comment]
Did you inject the mailerclient above in your class ?
Did you configured well application.conf?

**Comment by yuchaosydney** [comment]
yeah, I just put this block of code in my controller and it works, all configuration is correct and send emails without any problem.

I just want to get started with writing unit test in play-mailer with this very simple code.

**Comment by placebocurejava** [comment]
Unit test???
What does it mean!!

**Comment by yuchaosydney** [comment]
Sorry for confusing, I mean writing tests with MockitoSugar and PlaySpec etc in TDD practice.

Not sure I am in the right way or in the right place.

**Comment by placebocurejava** [comment]
I have no idea about these terms
Im happy that u finally send ur messages
If u want any help feel free to ask
FYI Im using play framework under java

**Comment by yuchaosydney** [comment]
Ok, then I will do R&D by myself, if got any new findings, I will let you guys know.

**Comment by wsargent** [comment]
Hi all -- placebocurejava has been banned from playframework, as he was posting on the google group with stolen credentials.  His comments have been removed.  Please see https://groups.google.com/forum/#!topic/play-framework/5FCht0abCI4 for more information.

**Comment by threeel** [comment]
@drhumlen maybe you can use something like this https://github.com/mailhog/MailHog have it as docker and use its API to do your test.
